eth-connector/src/lib.rs

```diff
@@ -707,9 +892,182 @@ mod tests {
+            assert_eq!(
+                eth_connector_balance_after_withdraw,
+                eth_connector_balance_before_withdraw + test_case.expected_fee_amount,
+                "Wrong eth-connector balance after withdraw"
+            );
+
+            owner_balance_before_withdraw = owner_balance_after_withdraw;
+            eth_connector_balance_before_withdraw = eth_connector_balance_after_withdraw;
+        }
+    }
+
+    #[test]
+    fn test_fee_on_withdraw() {
+        test_withdraw_generic(|contract, test_case| {
+            contract.withdraw(recipient_address(), test_case.amount)
+        });
+    }
+
+    #[test]
+    fn test_fee_on_engine_withdraw() {
+        test_withdraw_generic(|contract, test_case| {
+            contract.engine_withdraw(owner(), recipient_address(), test_case.amount)
+        });
+    }
+
+    #[test]
+    fn test_fee_on_finish_deposit() {
+        set_env!(predecessor_account_id: owner(), current_account_id: eth_connector(), attached_deposit: 1);
+        let mut contract = create_contract();
+
+        let mut owner_balance_before_deposit = contract.ft_balance_of(owner()).0;
+        assert_eq!(
+            owner_balance_before_deposit, 0,
+            "Wrong owner balance of the owner after mint"
+        );
+
+        let mut eth_connector_balance_before_deposit = contract.ft_balance_of(eth_connector()).0;
+        assert_eq!(
+            eth_connector_balance_before_deposit, 0,
+            "Wrong eth connector blanace after init"
+        );
+
+        for test_case in &get_fee_test_cases() {
+            contract.set_deposit_fee(Some(test_case.fee));
+            let proof_key = rand::distributions::DistString::sample_string(
+                &rand::distributions::Alphanumeric,
+                &mut rand::thread_rng(),
+                20,
+            );
+            contract.finish_deposit(
+                FinishDepositCallArgs {
+                    new_owner_id: owner(),
+                    amount: test_case.amount,
+                    proof_key,
+                    msg: None,
+                },
+                true,
+            );
+
+            let owner_balance_after_deposit = contract.ft_balance_of(owner()).0;
+            assert_eq!(
+                owner_balance_after_deposit,
+                owner_balance_before_deposit + test_case.expected_transferred_amount,
+                "Wrong owner balance after deposit"
+            );
+            let eth_connector_balance_after_deposit = contract.ft_balance_of(eth_connector()).0;
+            assert_eq!(
+                eth_connector_balance_after_deposit,
+                eth_connector_balance_before_deposit + test_case.expected_fee_amount,
+                "Wrong eth-connector balance after deposit"
+            );
+
+            owner_balance_before_deposit = owner_balance_after_deposit;
+            eth_connector_balance_before_deposit = eth_connector_balance_after_deposit;
+        }
+    }
```

Comment:
We also need tests for silos and tests when we set `None` as fee.

Reply:
c709963855d058e46d6d223fc2649cf7fde4156e